TACC/Core-CMS
# feat: add djangocms_tacc_image_gallery

## Overview

Add `djangocms_tacc_image_gallery` via https://github.com/wesleyboar/Core-CMS-Plugin-Image-Gallery.

## Related

- [TUP-318](https://jira.tacc.utexas.edu/browse/TUP-318)
- https://github.com/wesleyboar/Core-CMS-Plugin-Image-Gallery

## Changes

- **added** plugin to requirements
- **added** plugin to settings

## Testing

<details><summary>How to Create Gallery</summary>

1. on a page, add the plugin "Image Gallery"
1. within the plugin, verify only "Image" and Picture / Image" plugins can be added
1. within the plugin, add two or more plugin "Image" or Picture / Image"
1. verify images are automatically wrapped in an anchor tag
1. verify `<script>` to load `initialize.js` is in markup _only **once**_
1. verify `<link>` to load `gallery.css` is in markup _only **once**_

</details>

<details><summary>How to Use Gallery</summary>

1. click on an image in the gallery
1. use the LightGallery arrows to navigate through images
1. _either_ click outside the image _or_ press <kbd>esc</kbd> to leave LightGallery

</details>

<details><summary>How to Review Gallery</summary>

1. verify image gallery works (see "How to Use Gallery")
1. verify image grid has gaps between images and is responsive
1. verify images are automatically wrapped in an anchor tag
1. verify `<script>` to load `initialize.js` is in markup _only **once**_
1. verify `<link>` to load `gallery.css` is in markup _only **once**_

</details>

### on https://github.com/TACC/tup-ui

* https://dev.tup.tacc.utexas.edu/test/image-gallery/plugin/
* https://dev.tup.tacc.utexas.edu/news/latest-news/2022/09/24/stem-summer-camps-teach-data-skills-teamwork/

### via Django CMS QuickStart

1. Follow "Testing Steps" of https://github.com/wesleyboar/django-cms-quickstart/pull/5.

### via https://github.com/TACC/Core-CMS

* https://dev.cep.tacc.utexas.edu/test/core-cms-654/

## UI

| name | image
| - | -
| plugin added | ![plugin added](https://github.com/TACC/Core-CMS/assets/62723358/1ce6f1e4-6754-40fd-a033-9636f63aad4b)
| child plugin choices | ![child plugins limited](https://github.com/TACC/Core-CMS/assets/62723358/a9c50fa8-50dd-4a34-8d1f-8441f63316d6)
| child plugins added | ![child plugins added](https://github.com/TACC/Core-CMS/assets/62723358/f97eab14-dab8-4254-9dcb-b23fdbcafc88)
| gallery markup | ![div + a + img](https://github.com/TACC/Core-CMS/assets/62723358/709cd39c-acb2-4798-8ad1-f2f816e36424)
| gallery assets | ![link + script](https://github.com/TACC/Core-CMS/assets/62723358/af3e01bb-76a0-4a27-9d56-a99d714ec8e4)
| gallery interaction | ![gallery active](https://github.com/TACC/Core-CMS/assets/62723358/1a16ade0-8c8c-4477-8aec-7272b1215508)

## Notes

1. The plugin does **not** have database migrations.

2. An alternative solution (with drawbacks) is built at https://dev.tup.tacc.utexas.edu/test/image-gallery/.

    <details><summary>The drawbacks are disparate pieces and extra manual steps for each gallery.</summary>

    > **Note**
    > The extra steps per gallery are noted with a ➕.

    - create a Style plugin
		- ➕ add classes `s-image-grid`, `lightboxgallery`
	- create Image or Picture / Image plugins
	- ➕ wrap images in anchor markup _either_ via JavaScript or Link plugins
	- load LightGallery assets; do so only once regardless how many galleries are on a page
	- load styles for image grid and maybe for LightGallery custom UI

	</details>

	> **Note**
	> I think a plugin's cohesiveness is better than a list of manual or isolated or implicit actions.
